crate/operator/expand_volume.py

```diff
@@ -0,0 +1,525 @@
+
+    If we have specific master/hot/cold node types configured these would be
+    separate StatefulSets.
+    """
+    expected_number_of_nodes = 0
+    all_sts: V1StatefulSetList = await apps.list_namespaced_stateful_set(namespace)
+    for sts in all_sts.items:
+        if sts.metadata.labels["app.kubernetes.io/name"] == name:
+            expected_number_of_nodes += sts.spec.replicas
+
+    if expected_number_of_nodes == 0:
+        return
+
+    if not await is_cluster_healthy(conn_factory, expected_number_of_nodes, logger):
+        raise kopf.TemporaryError(
+            "Waiting for cluster to be healthy.", delay=config.HEALTH_CHECK_RETRY_DELAY
+        )
+
+
+class ExpandVolumeSubHandler(StateBasedSubHandler):
+    @crate.on.error(error_handler=crate.send_update_failed_notification)
+    @crate.timeout(timeout=90 if config.TESTING else config.EXPAND_VOLUME_TIMEOUT)
+    async def handle(  # type: ignore
+        self,
+        namespace: str,
+        name: str,
+        spec: kopf.Spec,
+        old: kopf.Body,
+        new: kopf.Body,
+        diff: kopf.Diff,
+        logger: logging.Logger,
+        **kwargs: Any,
+    ):
+        expand_data_diff_items: Optional[List[kopf.DiffItem]] = None
+
+        for operation, field_path, old_value, new_value in diff:
+            if field_path == ("spec", "nodes", "data"):
+                expand_data_diff_items = []
+                for node_spec_idx in range(len(old_value)):
+                    old_spec = old_value[node_spec_idx]
+                    new_spec = new_value[node_spec_idx]
+
+                    expand_data_diff_items.append(
+                        kopf.DiffItem(
+                            kopf.DiffOperation.CHANGE,
+                            (str(node_spec_idx), "resources", "disk", "size"),
+                            old_spec["resources"]["disk"]["size"],
+                            new_spec["resources"]["disk"]["size"],
+                        )
+                    )
+            else:
+                logger.info("Ignoring operation %s on field %s", operation, field_path)
+
+        if expand_data_diff_items:
+            async with ApiClient() as api_client:
+                core = CoreV1Api(api_client)
+
+                await expand_volume(
+                    core,
+                    namespace,
+                    name,
+                    kopf.Diff(expand_data_diff_items),
+                    logger,
+                )
+
+        # schedule success notification and send it after the cluster
+        # has been restarted successfully.
+        self.schedule_notification(
+            WebhookEvent.FEEDBACK,
+            WebhookFeedbackPayload(
+                message="The cluster storage has been resized successfully.",
+                operation=WebhookOperation.UPDATE,
+            ),
+            WebhookStatus.SUCCESS,
+        )
+
+
+class StartClusterSubHandler(StateBasedSubHandler):
```

Comment:
Do you think it might be useful to re-structure things a bit and move these generic handlers into a separate file? As they have nothing to do with expanding volumes per se...

Reply:
hm yes true. I think they would fit better into `operations.py`. I'll move them..